src/include/daos_types.h

```diff
@@ -828,44 +836,96 @@ enum {
+	DAOS_ACL_OWNER_GROUP,	/** Owning group */
+	DAOS_ACL_GROUP,		/** Group */
+	DAOS_ACL_EVERYONE	/** Anyone else */
+};
+
+/**
+ * Bits representing access types to set permissions for
+ */
+enum daos_acl_access_type {
+	DAOS_ACL_ACCESS_ALLOW = (1U << 0),	/** allow access */
+	DAOS_ACL_ACCESS_AUDIT = (1U << 1),	/** log the access for review */
+	DAOS_ACL_ACCESS_ALARM = (1U << 2)	/** notify of the access */
 };
 
-struct daos_acl_entry {
-	/** DAOS_ACL_USER/GROUP/MASK/OTHER */
-	uint16_t		ae_tag;
-	/** permissions DAOS_PC_RO/RW */
-	uint16_t		ae_perm;
-	/** uid or gid, meaningful only when ae_tag is DAOS_ACL_USER/GROUP */
-	uint32_t		ae_id;
+/**
+ * Bits representing access flags
+ */
+enum daos_acl_flags {
+	/** This represents a group, not a user */
+	DAOS_ACL_FLAG_GROUP		= (1U << 0),
+	/** Containers should inherit access controls from this pool */
+	DAOS_ACL_FLAG_POOL_INHERIT	= (1U << 1),
+	/** Audit/alarm should occur on failed access */
+	DAOS_ACL_FLAG_ACCESS_FAIL	= (1U << 2),
+	/** Audit/alarm should occur on successful access */
+	DAOS_ACL_FLAG_ACCESS_SUCCESS	= (1U << 3)
+};
+
+/**
+ * Bits representing the specific permissions that may be set
+ */
+enum daos_acl_perm {
+	DAOS_ACL_PERM_READ	= (1U << 0),
+	DAOS_ACL_PERM_WRITE	= (1U << 1)
 };
 
+/**
+ * Access Control Entry for a given principal.
+ * Each principal has at most one ACE that lists all their permissions in a
+ * given Access Control List.
+ */
+struct daos_ace {
+	/** Bitmap of daos_acl_access_type */
+	uint8_t		dae_access_types;
+	/** daos_acl_principal_type */
+	uint8_t		dae_principal_type;
+	/** Length of the principal string */
+	uint16_t	dae_principal_len;
+	/** Bitmap of daos_acl_flags */
+	uint16_t	dae_access_flags;
+	/** Reserved for 64-bit alignment */
+	uint16_t	dae_reserv;
+	/** Bitmap of daos_acl_perm for the ALLOW access */
+	uint64_t	dae_allow_perms;
+	/** Bitmap of daos_acl_perm for AUDIT access */
+	uint64_t	dae_audit_perms;
+	/** Bitmap of daos_acl_perm for ALARM access */
+	uint64_t	dae_alarm_perms;
+	/**
+	 * Null-terminated string representing the principal name for specific
+	 * user/group.
+	 * Actual bytes allocated MUST be rounded up for 64-bit alignment.
+	 * Empty for special principals OWNER, OWNER_GROUP, and EVERYONE.
+	 */
+	char		dae_principal[];
+};
+
+/**
+ * Header for the Access Control List, followed by the table of variable-length
+ * Access Control Entries.
+ * The entry list may be walked by inspecting the principal length and
+ * calculating the entry's
+ */
 struct daos_acl {
-	/** number of acl entries */
-	uint32_t		 da_nr;
-	/** reserved for future usage (for 64 bits alignment now) */
-	uint32_t		 da_reserv;
-	/** acl entries array */
-	struct daos_acl_entry	*da_entries;
+	/** Version of the table format */
+	uint16_t	dal_ver;
+	/** reserved for 64-bit alignment */
+	uint16_t	dal_reserv;
+	/** length of entries list in bytes */
+	uint32_t	dal_len;
+	/** table of variable-length Access Control Entries */
+	struct daos_ace	dal_ace[];
```

Comment:
As indexing will not be possible (right?), I wonder if it could not be simpler to define an array of char ?

Reply:
That's right, we can't actually index into this array. Agreed, an array of bytes probably would better communicate that.

Reply:
One thought came to mind after I pushed up the changes: Defining the ACE list as an array of `struct daos_ace` does give us one advantage. It lets us define a default ACL statically with a list of default ACEs. Not sure how high I'd prioritize that, since I can only think of one usage (initializing pool metadata), and defining statically doesn't save us much work in that case. @johannlombardi?